File: Graph algorithms/flows.py
class Edge:
    def __init__(self, to, w, f):
        self.to = to
        self.w = w
        self.f = f
       
def add_edge(fr, to, w):
    g[fr].append(len(e))
    e.append(Edge(to, w, 0))
    g[to].append(len(e))
    e.append(Edge(fr, 0, 0))

def rem(edge):
    return edge.w - edge.f

def dfs(v, f, k):
    if used[v]:
        return 0
    used[v] = True
    if v == T:
        return f
    for idx in g[v]:
        a = e[idx]
        r = rem(a)
        if r < k:
            continue
        pushed = dfs(a.to, min(r, f), k)
        if pushed:
            a.f += pushed
            e[idx ^ 1].f -= pushed
            return pushed
    return 0

S = 0
T = n - 1
ans = 0
for k in range(30, -1, -1):
    min_flow = 2 ** k
    while True:
        used = [False] * n
        flow = dfs(S, INF, min_flow)
        if not flow:
            break
        ans += flow
print(ans)

File: Graph algorithms/flows.cpp
struct edge {
    int to, w, f;
    edge(int to, int w, int f) {
        this->to = to;
        this->w = w;
        this->f = f;
    }
};
void add_edge(int fr, int to, int w) {
    g[fr].pb(e.size());
    e.pb(edge(to, w, 0));
    g[to].pb(e.size());
    e.pb(edge(fr, 0, 0));
}
int rem(edge x) {
    return x.w - x.f;
}
int dfs(int v, int f, int k, int T) {
    if (used[v]) return 0;
    used[v] = 1;
    if (v == T) return f;
    for (int i : g[v]) {
        edge &a = e[i];
        int r = rem(a);
        if (r < k) continue;
        int pushed = dfs(a.to, min(r, f), k, T);
        if (pushed) {
            a.f += pushed;
            e[i ^ 1].f -= pushed;
            return pushed;
        }
    }
    return 0;
}
void solve() {
    int ans = 0;
    for (int k = 30; k >= 0; k--) {
        int min_flow = 1 << k;
        while (true) {
            used.assign(n, 0);
            int flow = dfs(S, INF, min_flow, T);
            if (! flow) break;
            ans += flow;
        }
    }
    cout << ans << '\n';
}
